cli/command/swarm/join_token.go

```diff
@@ -12,73 +12,88 @@ import (
-	var rotate, quiet bool
+	opts := joinTokenOptions{}
 
 	cmd := &cobra.Command{
 		Use:   "join-token [OPTIONS] (worker|manager)",
 		Short: "Manage join tokens",
 		Args:  cli.ExactArgs(1),
 		RunE: func(cmd *cobra.Command, args []string) error {
-			worker := args[0] == "worker"
-			manager := args[0] == "manager"
-
-			if !worker && !manager {
-				return errors.New("unknown role " + args[0])
-			}
-
-			client := dockerCli.Client()
-			ctx := context.Background()
-
-			if rotate {
-				var flags swarm.UpdateFlags
-
-				swarm, err := client.SwarmInspect(ctx)
-				if err != nil {
-					return err
-				}
-
-				flags.RotateWorkerToken = worker
-				flags.RotateManagerToken = manager
-
-				err = client.SwarmUpdate(ctx, swarm.Version, swarm.Spec, flags)
-				if err != nil {
-					return err
-				}
-				if !quiet {
-					fmt.Fprintf(dockerCli.Out(), "Successfully rotated %s join token.\n\n", args[0])
-				}
-			}
-
-			swarm, err := client.SwarmInspect(ctx)
-			if err != nil {
-				return err
-			}
-
-			if quiet {
-				if worker {
-					fmt.Fprintln(dockerCli.Out(), swarm.JoinTokens.Worker)
-				} else {
-					fmt.Fprintln(dockerCli.Out(), swarm.JoinTokens.Manager)
-				}
-			} else {
-				info, err := client.Info(ctx)
-				if err != nil {
-					return err
-				}
-				return printJoinCommand(ctx, dockerCli, info.Swarm.NodeID, worker, manager)
-			}
-			return nil
+			opts.role = args[0]
+			return runJoinToken(dockerCli, opts)
 		},
 	}
 
 	flags := cmd.Flags()
-	flags.BoolVar(&rotate, flagRotate, false, "Rotate join token")
-	flags.BoolVarP(&quiet, flagQuiet, "q", false, "Only display token")
+	flags.BoolVar(&opts.rotate, flagRotate, false, "Rotate join token")
+	flags.BoolVarP(&opts.quiet, flagQuiet, "q", false, "Only display token")
 
 	return cmd
 }
 
+func runJoinToken(dockerCli *command.DockerCli, opts joinTokenOptions) error {
+	worker := opts.role == "worker"
+	manager := opts.role == "manager"
+
+	if !worker && !manager {
+		return errors.New("unknown role " + opts.role)
+	}
+
+	client := dockerCli.Client()
+	ctx := context.Background()
+
+	if opts.rotate {
+		flags := swarm.UpdateFlags{
+			RotateWorkerToken:  worker,
+			RotateManagerToken: manager,
+		}
+
+		sw, err := client.SwarmInspect(ctx)
+		if err != nil {
+			return err
+		}
+
+		if err := client.SwarmUpdate(ctx, sw.Version, sw.Spec, flags); err != nil {
+			return err
+		}
+
+		if !opts.quiet {
+			fmt.Fprintf(dockerCli.Out(), "Successfully rotated %s join token.\n\n", opts.role)
+		}
+	}
+
+	// second SwarmInspect in this function,
+	// this is necessary since SwarmUpdate after first changes the join tokens
+	sw, err := client.SwarmInspect(ctx)
+	if err != nil {
+		return err
+	}
+
+	if opts.quiet && worker {
+		fmt.Fprintln(dockerCli.Out(), sw.JoinTokens.Worker)
+		return nil
+	}
+
+	if opts.quiet && manager {
+		fmt.Fprintln(dockerCli.Out(), sw.JoinTokens.Manager)
+		return nil
+	}
+
+	info, err := client.Info(ctx)
+	if err != nil {
+		return err
+	}
```

Comment:
Actually, looking again; since we only use this info in the `printJoinCommand`, we can move this inside the function. Since `client.Info()` is a bit more heavyweight than `client.SwarmInspect()`, I suggest to move it after;

```
 	swarm, err := client.SwarmInspect(ctx)
 	if err != nil {
 		return err
 	}

	info, err := client.Info(ctx)
	if err != nil {
		return err
	}
```

the `nodeID` argument can be removed then from `printJoinCommand()`